consensus/src/common.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use blake2_rfc::blake2b::blake2b;
+use mina_crypto::hash::BaseHash;
+use mina_rs_base::consensus_state::ConsensusState;
+use mina_rs_base::global_slot::GlobalSlot;
+use mina_rs_base::protocol_state::{Header, ProtocolState};
+use serde_bin_prot::to_writer;
+use std::convert::TryInto;
+
+pub struct ProtocolStateChain(Vec<ProtocolState>);
+
+pub trait Chain<T>
+where
+    T: Header,
+{
+    fn push(&mut self, new: T) -> Result<(), &'static str>;
+    fn top(&self) -> Option<&T>;
+    fn consensus_state(&self) -> Option<&ConsensusState>;
+    fn global_slot(&self) -> Option<&GlobalSlot>;
+    fn epoch_slot(&self) -> Option<u32>;
+    fn length(&self) -> u64;
+    fn last_vrf(&self) -> String;
+    fn state_hash(&self) -> Option<BaseHash>;
+}
+
+impl Chain<ProtocolState> for ProtocolStateChain {
+    fn push(&mut self, new: ProtocolState) -> Result<(), &'static str> {
+        match self.0.len() {
+            0 => (),
+            n => {
+                if new.get_height().0 != self.0[n - 1].get_height().0 + 1 {
+                    return Err("header must have height 1 greater than top");
+                }
+            }
+        }
+
+        self.0.push(new);
+        Ok(())
+    }
+
+    fn top(&self) -> Option<&ProtocolState> {
+        self.0.last()
+    }
+
+    fn consensus_state(&self) -> Option<&ConsensusState> {
+        match self.top() {
+            Some(s) => Some(&s.body.consensus_state),
+            None => None,
+        }
+    }
+
+    fn global_slot(&self) -> Option<&GlobalSlot> {
+        match self.top() {
+            Some(s) => Some(&s.body.consensus_state.curr_global_slot),
+            None => None,
+        }
+    }
+
+    fn epoch_slot(&self) -> Option<u32> {
+        self.global_slot()
+            .map(|s| (s.slot_number.0 % s.slots_per_epoch.0))
+    }
+
+    fn length(&self) -> u64 {
+        self.0.len().try_into().unwrap()
+    }
+
+    fn last_vrf(&self) -> String {
+        let s = match self.top() {
+            Some(s) => s,
+            None => return "0x".to_string(),
+        };
+
+        let hash = blake2b(
+            32,
+            &[],
+            &s.body.consensus_state.last_vrf_output.0.as_bytes(),
+        );
+        BaseHash::from(hash.as_bytes()).to_hex()
+    }
+
+    fn state_hash(&self) -> Option<BaseHash> {
```

Comment:
It is important to use the correct hash types. By convention Mina adds a prefix string to data before hashing it which is different depending on the type being hashes. Presumably this is to prevent a hash collision between different types with identical data. 

Check out the prefixes here https://github.com/ChainSafe/mina-rs/blob/8daf8e2a45e19c3b4912c688744faaa20cee6388/crypto/src/hash/prefixes.rs#L76

To avoid confusion I made the `Hash` and `Hashable` traits. Here `ProtocolState` should implement `Hashable` which returns a `StateHash` which implements `Hash`. This also makes these parts of the code much tidier since all the hashing logic will live in once place.

This is a bit tricky to describe in a comment so I'll link a branch showing how this could be done using our hashing lib